File: frontend/app/src/main/java/com/h1oo7/musicapp/network/RetrofitClient.java
// File: app/java/com/h1oo7/musicapp/network/RetrofitClient.java
package com.h1oo7.musicapp.network;

import com.h1oo7.musicapp.App;
import com.h1oo7.musicapp.utils.Constants;
import com.h1oo7.musicapp.utils.SharedPrefManager;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitClient {
    private static Retrofit retrofit = null;

    public static Retrofit getClient() {
        if (retrofit == null) {
            // In log để debug
            HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
            logging.setLevel(HttpLoggingInterceptor.Level.BODY);

            // Tự động thêm token vào header
            OkHttpClient client = new OkHttpClient.Builder()
                    .addInterceptor(chain -> {
                        Request original = chain.request();
                        String token = SharedPrefManager.getInstance(App.getContext()).getToken();

                        Request.Builder builder = original.newBuilder();
                        if (token != null) {
                            builder.addHeader("Authorization", "Bearer " + token);
                        }
                        builder.addHeader("Content-Type", "application/json");

                        return chain.proceed(builder.build());
                    })
                    .addInterceptor(logging)
                    .build();

            retrofit = new Retrofit.Builder()
                    .baseUrl(Constants.BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .client(client)
                    .build();
        }
        return retrofit;
    }
}

File: forntend/app/src/main/java/com/example/musique/data/network/RetrofitClient.kt
package com.example.musique.data.network

import com.example.musique.utils.PreferenceManager
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit
import com.example.musique.utils.Constants


object RetrofitClient {
    
//    private const val BASE_URL = "http://10.0.2.2:5000/api/"
//    for real phone
    private const val BASE_URL = "http://192.168.0.103:5000/api/"


    private var retrofit: Retrofit? = null
    
    fun getClient(preferenceManager: PreferenceManager): Retrofit {
        if (retrofit == null) {
            val loggingInterceptor = HttpLoggingInterceptor().apply {
                level = HttpLoggingInterceptor.Level.BODY
            }
            
            val client = OkHttpClient.Builder()
                .addInterceptor(AuthInterceptor(preferenceManager))
                .addInterceptor(loggingInterceptor)
                .connectTimeout(Constants.NETWORK_TIMEOUT_SECONDS, TimeUnit.SECONDS)
                .readTimeout(Constants.NETWORK_TIMEOUT_SECONDS, TimeUnit.SECONDS)
                .build()
            
            retrofit = Retrofit.Builder()
                .baseUrl(Constants.API_BASE_URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
        }
        return retrofit!!
    }
    
    fun getApiService(preferenceManager: PreferenceManager): ApiService {
        return getClient(preferenceManager).create(ApiService::class.java)
    }
}
